### DISCUSSION TITLE
Buffer follow behavior inqiry.

### DISCUSSION THREAD

**Discussion by YourDadsHusband** [chimay/wheel#3] [Q&A]
I am attempting to create a statusline displaying the torus and wheel information associated with the currently open buffer, and am experiencing issues with the buffer follow behavior. Navigating through wheel utilities behaves as expected and updates the information correctly, however if I open a buffer using :e or switch by specifying a buffer to jump to through vim the torus/wheel info is desynced. At this point neither manual calls of wheel#projection#follow or using wheel navigation will restore the correct information. I feel as if there is something I am missing or misunderstanding about the functions to update the wheel tracking information. As an aside, is there a clean, nondestructive way to clear the current followed file without unloading the entire wheel?

**Comment by chimay** [comment]
The follow routine should detect that you are editing a file belonging to the wheel.

Have you activated an autocommand to  launch the follow function on bufenter, like this :

```
    autocmd BufEnter * call wheel#projection#follow()
```

Calling manually the function should do the work, anyway, strange. Maybe with your wheel config I could investigate further.

Edit : I assume the file you are editing is belonging to the wheel already ? In the contrary, wheel#projection#follow will do nothing.

If by clearing the current file you mean remove it from the wheel, it’s simply `<M-w>d`, but maybe I misunterstood your question ?

**Reply by YourDadsHusband** [reply]
I have in fact set up the autocommand on buffer entry, as well as one that calls it on displaying an already open buffer in a window. and that does behave normally in most circumstances. It seems to me that the desynchronization occurs most often when I am switching between two files. I have a keybind set up to call :b# to rapidly switch back to the most recently opened buffer, and that is the most common action that causes wheel to lose track of the currently open file. 

My current wheel config is the default example configuration, the only changes being my local pathnames. I would appreciate if you would take a look at my code for what I am actually trying to impliment, which is a statusbar display for the current torus/circle. Perhaps the behavior I am experienceing is a side effect from improperly fetching the values I want?

`function! WheelStatus ()

    "Wheel statusbar widget
    "Call follow script to update location
    call wheel#projection#follow()

    let [torus, circle, location] = wheel#referen#location('all')
    if ! wheel#referen#empty('wheel')
        "Fetch torus name or identify if it is empty
        let Torus = torus.name
    else
        let Torus = "No Torus Found"
    endif
    if ! wheel#referen#empty('torus')
        "Fetch circle name or identify if it is empty
        let Circle = circle.name
    else
        let Circle = "No Circle Found"
    endif
    let WheelStatus=" [".Torus." > ".Circle."]"
    return WheelStatus
endfunction
`

I was actually asking if there is a proper way to set just the value for the current Torus, Wheel, and Location to null without modifying any other associated information or markers.

**Reply by chimay** [reply]
Strange, I was sure I checked for your answer, but missed it, sorry.

If you alternate with a non-wheel file, the current state will remain as the last visited file in the wheel, 
and your status will not be updated. So, you could check in your code if you are in the wheel, or not. You can
use the `wheel#referen#is_in_wheel` function, I just wrote it and pushed, so don't forget to pull if you want to use it.

Here is an example:

```
function! WheelStatus ()

"Wheel statusbar widget

if ! wheel#referen#is_in_wheel ()
" not in wheel, let's print a special status, e.g.:
let WheelStatus =  "[Outside of the wheel galaxy]"
return WheelStatus
endif

" Now, we are in a wheel file, normal status

"Call follow script to update location
call wheel#projection#follow()

let [torus, circle, location] = wheel#referen#location('all')
if ! wheel#referen#empty('wheel')
    "Fetch torus name or identify if it is empty
    let Torus = torus.name
else
    let Torus = "No Torus Found"
endif
if ! wheel#referen#empty('torus')
    "Fetch circle name or identify if it is empty
    let Circle = circle.name
else
    let Circle = "No Circle Found"
endif
let WheelStatus=" [".Torus." > ".Circle."]"
return WheelStatus
endfunction
```

The is_in_wheel function:

```
fun! wheel#referen#is_in_wheel (...)
	" Whether filename argument is in wheel
	" Default optional argument : current filename
	if a:0 > 0
		let filename = a:1
	else
		let filename = expand('%:p')
	endif
	let wheel_files = wheel#helix#files ()
	let is_in_wheel = index(wheel_files, filename) >= 0
	return is_in_wheel
endfun
```
